[WIP] multiplexer config file

jackloughran opened the pull request:
@Stebalien is this along the lines of what you were thinking for #4893?

the `--enable-mplex-experiment` flag still exists, but it will be ignored if there are multiplexers specified in the config file. i also made the configuration options simply `yamax` and `mplex` because it looks like the full protocol name was hard coded anyway.

i will update the documentation and command help information if this approach is the direction you want to go in.

thanks!

Stebalien commented on core/core.go:
```diff
@@ -382,25 +382,46 @@ func makeAddrsFactory(cfg config.Addresses) (p2pbhost.AddrsFactory, error) {
+		} else if multiplexer == "mplex" {
+			isMplex = true
+		}
 	}
```
We should at least log if we find a multiplexer we don't know about.

Stebalien commented on core/core.go:
```diff
@@ -382,25 +382,46 @@ func makeAddrsFactory(cfg config.Addresses) (p2pbhost.AddrsFactory, error) {
+	}
 
-	if mplexExp {
+	if isMplex || mplexExp {
```
It would be nice to warn if `--enable-mplex-experiment` is `false` but mplex is explicitly enabled in the config.

Stebalien commented on core/core.go:
```diff
@@ -382,33 +382,76 @@ func makeAddrsFactory(cfg config.Addresses) (p2pbhost.AddrsFactory, error) {
+	// Allow muxer preference order overriding.  If set, LIBP2P_MUX_PREFS will
+	// overwrite the order specified in the config.
 	if prefs := os.Getenv("LIBP2P_MUX_PREFS"); prefs != "" {
 		mstpt.OrderPreference = strings.Fields(prefs)
```
Nit: can we add a `log.Error(...)` here warning the user that `LIBP2P_MUX_PREFS` is deprecated (telling them to use `Swarm.Multiplexers`).

Stebalien commented on core/core.go:
```diff
@@ -382,33 +382,76 @@ func makeAddrsFactory(cfg config.Addresses) (p2pbhost.AddrsFactory, error) {
 		mstpt.OrderPreference = strings.Fields(prefs)
 	}
 
+	if isYamex {
```
This needs to go before setting the order preference (will cause problems otherwise).

Stebalien commented on core/core.go:
```diff
@@ -382,33 +382,76 @@ func makeAddrsFactory(cfg config.Addresses) (p2pbhost.AddrsFactory, error) {
+					"but mplex is given as a multiplexer in the config.")
+			}
+		} else {
+			log.Warningf("Unknown multiplexer specified in config: %s", multiplexer)
```
nit: I'd turn these warnings into errors. That way, they'll get printed to stdout on daemon start (we'd like to print warnings as well but they're too noisy at the moment).

jackloughran commented:
@Stebalien thanks for all the feedback.  i changed the logging to `log.Error` and i fixed the order of the statements.  it looks like i need to add some more testing to cover some of these new statements i added.  i also still have to update the documentation to reflect this new configuration option.  are there any changes you think i should make before i do that stuff?

Kubuxu commented on core/core.go:
```diff
@@ -382,31 +382,76 @@ func makeAddrsFactory(cfg config.Addresses) (p2pbhost.AddrsFactory, error) {
-		MaxStreamWindowSize:    uint32(1024 * 512),
-		LogOutput:              ioutil.Discard,
+	multiplexerProtocolNames := map[string]string{
+		"yamex": "/yamux/1.0.0",
```
Any reason for it being called `yamex` instead of `yamux`?

jackloughran replied:
oh wow thanks for catching that.  im not sure why i did that (i also spelled it with an a above 😬).  ill fix this right now.

Kubuxu commented:
Hmm, I am sure I sent out another comment with the one about yamux but it didn't go through, writing it now.

Kubuxu commented on core/core.go:
```diff
@@ -382,31 +382,76 @@ func makeAddrsFactory(cfg config.Addresses) (p2pbhost.AddrsFactory, error) {
+	var isYamux, isMplex bool
+
+	for _, multiplexer := range multiplexers {
+		if multiplexer == "yamux" {
```
I think it will be better to use case statement here instead of ifs.

Kubuxu commented on core/core.go:
```diff
@@ -382,31 +382,76 @@ func makeAddrsFactory(cfg config.Addresses) (p2pbhost.AddrsFactory, error) {
+
+	// If there are any multiplexers defined in the config, ignore mplexExp.
+	if len(multiplexers) > 0 {
+		mplexExp = false
```
Is there any reason to set it to false, it doesn't seem to be used after.

Kubuxu commented on core/core.go:
```diff
@@ -382,31 +382,76 @@ func makeAddrsFactory(cfg config.Addresses) (p2pbhost.AddrsFactory, error) {
+	for _, multiplexer := range multiplexers {
+		if multiplexer == "yamux" {
+			isYamux = true
+		} else if multiplexer == "mplex" {
```
@Stebalien @whyrusleeping do you think using just protocol names is good enough here or should we use full protocol descriptor. This way if we ever want to have mplex6 and mplex7 (if it ever exists) both at the same time, it will be possible.

Stebalien replied:
I don't have much of a preference.

jackloughran replied:
@Stebalien @Kubuxu sorry for letting this get stale. i can definitely update the pr to use full protocol descriptors if you guys think that's better.

Kubuxu replied:
I don't think this is necessary. We can just have `mplex` and in future add others.
